src/DayDifference.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace johncorrelli;
+
+use DateTimeInterface;
+
+class DayDifference
+{
+    /**
+     * Number of seconds in one day.
+     */
+    const ONE_DAY = 86400;
+
+    /**
+     * @var array<int> 0-6 representation of days of the week, where 0 = Sunday, and 6 = Saturday
+     */
+    private array $allowedDaysOfTheWeek = [];
+
+    private DateTimeInterface $endDate;
+
+    /**
+     * @var array<string> YYYY-MM-DD Dates that should not be counted. ie: 2020-01-01 or *-07-04 for dates that repeat each year.
+     */
+    private array $excludedDates = [];
+
+    private DateTimeInterface $startDate;
+
+    /**
+     * @param array<int>    $allowedDaysOfTheWeek
+     * @param array<string> $excludedDates
+     */
+    public function __construct(
+        DateTimeInterface $startDate,
+        DateTimeInterface $endDate,
+        array $allowedDaysOfTheWeek = [0, 1, 2, 3, 4, 5, 6],
+        array $excludedDates = []
+    ) {
+        $this->startDate = $startDate;
+        $this->endDate = $endDate;
+        $this->allowedDaysOfTheWeek = $allowedDaysOfTheWeek;
+        $this->excludedDates = $excludedDates;
+    }
+
+    /**
+     * Determines the number of allowed days between a start and an end date.
+     */
+    public function difference(): int
+    {
+        $isPositive = $this->startDate <= $this->endDate;
+
+        if ($isPositive) {
+            $multiply = 1;
+            $startStamp = $this->startDate->getTimestamp();
+            $endStamp = $this->endDate->getTimestamp();
+        } else {
+            $multiply = -1;
+            $startStamp = $this->endDate->getTimestamp();
+            $endStamp = $this->startDate->getTimestamp();
+        }
+
+        return $this->findDifference($startStamp, $endStamp, $this->allowedDaysOfTheWeek, $this->excludedDates) * $multiply;
+    }
+
+    /**
+     * Returns the number of allowed days between the start and end timestamp.
+     *
+     * @param array<int>    $allowedDaysOfTheWeek
+     * @param array<string> $excludedDates
+     */
+    private function findDifference(int $startStamp, int $endStamp, array $allowedDaysOfTheWeek = [], array $excludedDates = []): int
+    {
+        $isFullWeek = count($allowedDaysOfTheWeek) === 7;
+        $hasExclusions = !empty($excludedDates);
+        $currentStamp = $startStamp;
+        $totalDays = 0;
+
+        while ($currentStamp < $endStamp) {
+            $totalDays += (int) (
+                ($isFullWeek || $this->isAllowed($currentStamp, $allowedDaysOfTheWeek))
```

Comment:
Is this converting a bool to an int? Is that the part that is faster than the `var++`? 

What do you think about breaking this out into a variable like `$shouldIncrement` or something so the lines would read easier?

Reply:
Good question... I spent a lot of time on what these few lines are doing....

Some background:
```
$a = (int) true; // = 1
$b = (int) false; // = 0
```

Essentially we're saying `$totalDays += 1` if the day is counted and `$totalDays += 0` if the day is excluded.

I initially had this as a check like you're suggesting, but it's surprising how much faster (on a large data set) not doing a variable assignment and an `if (should increment) { $totalDays++; }` is.

Perhaps I should add some comments in code to explain why?

Reply:
Ah, that makes sense. I don't think you need comments, but if you renamed the variables or grouped them differently so it was more indicative of what that is doing, that could be helpful. 

I wasn't necessarily suggesting breaking it out beforehand, but just some thing like this...

```php
$isDayOfWeekAllowed =  ($isFullWeek || $this->isAllowed($currentStamp, $allowedDaysOfTheWeek));
$isDateAllowed = (!$hasExclusions || !$this->isExcluded($currentStamp, $excludedDates));

$incrementByOneOrZero = (int) ($isDayOfWeekAllowed && $isDateAllowed);

$totalDays += incrementByOneOrZero;
```

Reply:
This is wild... dropping that change into the test case on TG makes the response time jump pretty significantly....

![Screenshot_20201012_101122](https://user-images.githubusercontent.com/18272064/95756589-f5d3c180-0c73-11eb-9ae8-f0f5edc906e2.png)

I did a multiple refreshes and the slowest was pretty much always between 12 and 14.

I think we're going to want to leave what we have... I added some comments to show whats up.